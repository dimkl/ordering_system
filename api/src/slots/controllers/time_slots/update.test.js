/**
 * @integration-test true
 * @data-factory true
 */
describe("PATCH /time_slots/:time_slot_id", () => {
  let knex;
  beforeAll(() => knex = require('../../../shared/setupModels')());
  beforeEach(() => knex.raw('truncate customers, users, time_slots, slots, sections, shops cascade;'));
  afterAll(() => knex.destroy());

  it("updates and returns updated time_slot", async () => {
    const timeSlot = await DataFactory.createTimeSlot();
    const slot = await DataFactory.createSlot({}, {}, { email: 'aloha@example.com' });
    const customer = await DataFactory.createCustomer({ email: 'timeSlot.update@example.com' });

    const response = await request.patch(`/time_slots/${timeSlot.uuid}`)
      .send({
        started_at: '2022-04-21T15:28:27.602Z',
        ended_at: '2022-05-21T15:28:27.602Z',
        customer_id: customer.uuid,
        slot_id: slot.uuid
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body).toMatchObject({
      "started_at": '2022-04-21T15:28:27.602Z',
      "ended_at": '2022-05-21T15:28:27.602Z',
      "created_at": timeSlot.created_at.toISOString(),
      "updated_at": expect.any(String),
      "id": timeSlot.id,
      "customer_id": customer.uuid,
      "slot_id": slot.uuid,
      "uuid": timeSlot.uuid
    });
    expect(timeSlot.customer_id).not.toEqual(customer.id);
    expect(timeSlot.slot_id).not.toEqual(slot.id);
    expect(timeSlot.started_at).not.toEqual('2022-04-21T15:28:27.602Z');
    expect(timeSlot.ended_at).not.toEqual('2022-05-21T15:28:27.602Z');
    const unixUpdatedAt = new Date(response.body.updated_at).getTime();
    expect(unixUpdatedAt).toBeGreaterThan(timeSlot.updated_at.getTime());
  });

  it("does not have required properties", async () => {
    const timeSlot = await DataFactory.createTimeSlot();

    const response = await request.patch(`/time_slots/${timeSlot.uuid}`)
      .send({})
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body).toMatchObject({
      "started_at": timeSlot.started_at.toISOString(),
      "ended_at": null,
      "created_at": timeSlot.created_at.toISOString(),
      "updated_at": timeSlot.updated_at.toISOString(),
      "id": timeSlot.id,
      "customer_id": timeSlot.customer.uuid,
      "slot_id": timeSlot.slot.uuid,
      "uuid": timeSlot.uuid
    });
  });

  it("throws validation error with additional properties", async () => {
    const timeSlot = await DataFactory.createTimeSlot();

    const response = await request.patch(`/time_slots/${timeSlot.uuid}`)
      .send({ created_at: Date.now() })
      .set("Accept", "application/json");

    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
  });
});