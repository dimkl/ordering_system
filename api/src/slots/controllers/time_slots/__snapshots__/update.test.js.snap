// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PATCH /time_slots/:time_slot_id throws validation error with additional properties 1`] = `
Array [
  Object {
    "instancePath": "",
    "keyword": "additionalProperties",
    "message": "must NOT have additional properties",
    "params": Object {
      "additionalProperty": "created_at",
    },
    "schemaPath": "#/additionalProperties",
  },
]
`;
